velox/exec/tests/WindowTest.cpp

```diff
@@ -79,6 +81,196 @@ TEST_F(WindowTest, spill) {
   ASSERT_GT(stats.spilledPartitions, 0);
 }
 
+TEST_F(WindowTest, rowBasedStreamingWindowMemoryUsageLower) {
+  const vector_size_t size = 100'000;
+  auto data = makeRowVector(
+      {"d", "p", "s"},
+      {
+          // Payload.
+          makeFlatVector<int64_t>(size, [](auto row) { return row; }),
+          // Partition key.
+          makeFlatVector<int16_t>(size, [](auto row) { return row % 11; }),
+          // Sorting key.
+          makeFlatVector<int32_t>(size, [](auto row) { return row; }),
+      });
+
+  createDuckDbTable({data});
+
+  core::PlanNodeId windowId;
+  auto plan = PlanBuilder()
+                  .values(split(data, 10))
+                  .window({"row_number() over (partition by p order by s)"})
+                  .capturePlanNodeId(windowId)
+                  .planNode();
+
+  auto task =
+      AssertQueryBuilder(plan, duckDbQueryRunner_)
+          .config(core::QueryConfig::kPreferredOutputBatchBytes, "1024")
+          .assertResults(
+              "SELECT *, row_number() over (partition by p order by s) FROM tmp");
+
+  auto taskStats = exec::toPlanStats(task->taskStats());
+  const auto& sortWindowStats = taskStats.at(windowId);
+
+  plan = PlanBuilder()
+             .values(split(data, 10))
+             .orderBy({"p", "s"}, false)
+             .streamingWindow({"row_number() over (partition by p order by s)"})
+             .capturePlanNodeId(windowId)
+             .planNode();
+
+  task =
+      AssertQueryBuilder(plan, duckDbQueryRunner_)
+          .config(core::QueryConfig::kPreferredOutputBatchBytes, "1024")
+          .assertResults(
+              "SELECT *, row_number() over (partition by p order by s) FROM tmp");
+
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::Window::supportRowsStreaming",
+      std::function<void(bool*)>([&](bool* supportRowsStreamingWindow) {
+        ASSERT_EQ(*supportRowsStreamingWindow, true);
+      }));
+
+  taskStats = exec::toPlanStats(task->taskStats());
+  const auto& streamingWindowStats = taskStats.at(windowId);
+
+  // sortWindowStats.peakMemoryBytes: 2418624
+  // streamingWindowStats.peakMemoryBytes: 124672
+  ASSERT_EQ(
+      sortWindowStats.peakMemoryBytes / streamingWindowStats.peakMemoryBytes,
```

Comment:
It would be great to have more insight into how 19 has come by.

sortWindowStats -> requires memory proportional to the data size.
streamingWindowStats -> Since there is no repetition of sort values within a partition the size is proportional to an input batch.

Instead of using 19 directly, might be worth comparing with the results of some formula that uses the characteristics of the data.

Reply:
@aditi-pandit I have removed the number 19 here because the memory usage does not increase linearly, making it difficult to represent with a formula. My current tests show that the memory usage of sort-based windows varies between different datasets, whereas the memory usage of Row-based windows tends to stabilize.